btree/btree.c

```diff
@@ -5358,6 +5470,215 @@ static void m0_btree_ut_mt_mt_kv_oper(void)
+		 */
+
+		M0_BTREE_OP_SYNC_WITH_RC(&b_op.bo_op,
+					 m0_btree_create(temp_node, 1024,
+							 &btree_type, nt, tx,
+							 &b_op),
+					 &b_op.bo_sm_group, &b_op.bo_op_exec);
+
+		tree = b_op.bo_arbor;
+
+		random_r(&ti->ti_random_buf, &rec_count);
+		rec_count %= MAX_RECS_FOR_TREE_TEST;
+		rec_count = rec_count ? : (MAX_RECS_FOR_TREE_TEST / 2);
+
+		for (i = 1; i <= rec_count; i++) {
+			key   = i;
+			value = key;
+
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+						 m0_btree_put(tree, tx,
+							      &rec, &ut_cb,
+							      0, &kv_op),
+						 &kv_op.bo_sm_group,
+						 &kv_op.bo_op_exec);
+			M0_ASSERT(data.flags == M0_BSC_SUCCESS);
+		}
+
+		m0_btree_close(tree);
+
+		m0_btree_open(temp_node, 1024, &tree);
+
+		for (i = 1; i <= rec_count; i++) {
+			key   = i;
+			value = key;
+
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+						 m0_btree_get(tree, &rec.r_key,
+							      &ut_cb, 0,
+							      &kv_op),
+						 &kv_op.bo_sm_group,
+						 &kv_op.bo_op_exec);
+			M0_ASSERT(data.flags == M0_BSC_SUCCESS);
+		}
+
+		m0_btree_close(tree);
+
+		m0_btree_open(temp_node, 1024, &tree);
+
+		for (i = 1; i <= rec_count; i++) {
+			key   = i;
+			value = key;
+
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+						 m0_btree_del(tree, &rec.r_key,
+							      &ut_cb, 0,
+							      &kv_op),
+						 &kv_op.bo_sm_group,
+						 &kv_op.bo_op_exec);
+			M0_ASSERT(data.flags == M0_BSC_SUCCESS);
+		}
+
+		m0_btree_close(tree);
+
+		M0_BTREE_OP_SYNC_WITH_RC(&b_op.bo_op,
+					 m0_btree_destroy(tree, &b_op),
+					 &b_op.bo_sm_group, &b_op.bo_op_exec);
+	}
+
+	m0_free_aligned(temp_node, (1024 + sizeof(struct nd)), 10);
+}
+
+static void btree_ut_num_threads_tree_oper(uint32_t thread_count)
+{
+	uint16_t                    *cpuid_ptr;
+	uint16_t                     cpu_count;
+	size_t                       cpu_max;
+	struct btree_ut_thread_info *ti;
+	uint16_t                     cpu;
+	int                          i;
+	int                          rc;
+
+	btree_ut_init();
+	online_cpu_id_get(&cpuid_ptr, &cpu_count);
+
+	if (thread_count == 0)
+		thread_count = cpu_count - 1; /** Skip Core-0 */
+
+	thread_start = false;
+
+	M0_ALLOC_ARR(ti, thread_count);
+	M0_ASSERT(ti != NULL);
+
+	cpu_max = m0_processor_nr_max();
+
+	cpu = 1; /** We skip Core-0 for Linux kernel and other processes. */
+	for (i = 0; i < thread_count; i++) {
+		rc = m0_bitmap_init(&ti[i].ti_cpu_map, cpu_max);
+		m0_bitmap_set(&ti[i].ti_cpu_map, cpuid_ptr[cpu], true);
+		cpu++;
+		if (cpu >= cpu_count)
+			/**
+			 *  Circle around if thread count is higher than the
+			 *  CPU cores in the system.
+			 */
+			cpu = 1;
+
+		ti[i].ti_thread_id  = i;
+	}
+
+	for (i = 0; i < thread_count; i++) {
+		rc = M0_THREAD_INIT(&ti[i].ti_q, struct btree_ut_thread_info *,
+				    btree_ut_thread_init,
+				    &btree_ut_tree_oper_thread_handler, &ti[i],
+				    "Thread-%d", i);
+		M0_ASSERT(rc == 0);
+	}
+
+	/** Initialized all the threads. Now start the chaos ... */
+	thread_start = true;
```

Comment:
Do you want to wait in the handler till all the threads are started? If yes, please add while check in handler or else you can remove this flag.

Reply:
There is no need to wait because when we come out of M0_THREAD_INIT the thread has already started running and it is waiting on this variable. If there are multiple threads (like we have in the loop) then all the threads are started and are waiting on this variable to change to 'true'.